File: JavaBaseWrapper/src/main/java/eu/fau/cs7/daceDS/Component/Consumer.java
/*******************************************************************************
 * Copyright 2021 Moritz Gütlein
 * 
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not
 * use this file except in compliance with the License.  You may obtain a copy
 * of the License at
 * 
 *   http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
 * License for the specific language governing permissions and limitations under
 * the License.
 ******************************************************************************/
package eu.fau.cs7.daceDS.Component;
import java.util.ArrayList;
import java.util.List;
import java.util.regex.Pattern;
import java.util.regex.Matcher;

import java.util.HashMap;


/** 
 * A generic consumer class that lets a callback function process messages 
 * that are received.
 *  
 * @author guetlein
 *
 * @param <T>
 */
public abstract class Consumer<T>
{
	protected String id;
	protected static final int DEFAULT_TIMEOUT = 50;    
	protected ConsumerCallback cb;
	protected Thread t;	
	protected boolean countingConsumer = true;
	protected boolean isListening;
	protected static HashMap<String, Long> receivedCounting = new HashMap<String, Long>();
	protected static List<String> subscribedTopics = new ArrayList<String>();
	protected static List<String> countingSubscribedTopics = new ArrayList<String>();
	protected static List<Pattern> subscribedPatterns = new ArrayList<Pattern>();//Pattern.compile("undefined");
	protected static List<Pattern> countingSubscribedPatterns = new ArrayList<Pattern>();//Pattern.compile("undefined");

	public Consumer(String id) {	this.id = id;};

	public Consumer(String id, boolean consuming) {		this.countingConsumer = consuming;
	this.id = id;};

	public  abstract void init(List<String> topics, ConsumerCallback callback);
	
	public abstract  void initPlainString(List<String> topics, ConsumerCallback callback);
	
	public abstract  void init(List<String> topics, ConsumerCallback callback, boolean plainString);

	public  abstract void initPattern(String pattern, ConsumerCallback callback);
	
	public  abstract void initPattern(String pattern, ConsumerCallback callback, boolean fetchLatest, boolean autopoll);

	public  abstract void init(List<String> topics, ConsumerCallback callback,  boolean fetchLatest, boolean autopoll, boolean plainString);

	public abstract void poll();

	public  abstract void poll(int timeout);
	
	public  abstract void close();


	//////////////////////////////////
	//counting

	public static List<String> getSubscribedTopics() {
		return subscribedTopics;
	}
	public static List<String> getCountingSubscribedTopics() {
		return countingSubscribedTopics;
	}

	public static String getSubscribedPattern() {
		String pstr = "";
		for(Pattern p : subscribedPatterns) {
			pstr+=p.pattern()+",";
		}
		return pstr;
	}
	public static String getCountingSubscribedPattern() {
		String pstr = "";
		for(Pattern p : countingSubscribedPatterns) {
			pstr+=p.pattern()+",";
		}
		return pstr;
	}

	public static HashMap<String, Long> getReceivedCounting(){
		return receivedCounting;
	}

	public static boolean hasSubscribed(String topic) {
		boolean t = subscribedTopics.contains(topic);
		if (t) return true;

		for(Pattern p : subscribedPatterns) {
			Matcher m = p.matcher(topic);
			if (m.matches()) 
				return true;
		}
		
		return false;
	}

	public static boolean hasSubscribedCounting(String topic) {
		boolean t = countingSubscribedTopics.contains(topic);
		if (t) return true;

		for(Pattern p : countingSubscribedPatterns) {
			Matcher m = p.matcher(topic);
			if (m.matches()) 
				return true;
		}

		return false;
	}

}

File: CppBaseWrapper/src/communication/Consumer.cpp
/*******************************************************************************
 * Copyright 2021 Moritz Gütlein
 * 
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not
 * use this file except in compliance with the License.  You may obtain a copy
 * of the License at
 * 
 *   http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
 * License for the specific language governing permissions and limitations under
 * the License.
 ******************************************************************************/
#include "Consumer.h"
using namespace daceDS;

std::map<std::string, int64_t> Consumer::countingReceiveCounter;
std::vector<std::string> Consumer::countingExpectedTopics;
std::vector<std::string> Consumer::countingExpectedPatterns;

Consumer::Consumer(std::shared_ptr<MessageHandler> hdl, bool isCounting = false) {
    handler = hdl;
    counting = isCounting;
};

void Consumer::subscribe(std::string broker,
                         std::vector<std::string> &topics,
                         std::string group_id,
                         std::string loglevel) {
    if (!isCounting()) {
        return;
    }
    for (std::string topic : topics) {
        countingReceiveCounter[topic] = 0;
        if (topic[0] == '^') {
            KDBGCB("added " << topic << " to the list of expected patterns");
            Consumer::countingExpectedPatterns.push_back(topic);
        } else {
            KDBGCB("added " << topic << " to the list of expected topics");
            Consumer::countingExpectedTopics.push_back(topic);  //todo: delete this list and set countingreceivecounter[t]=0?
        }
    }
}

bool Consumer::process(ConsumedMessage* cmsg) {

    // if we are not counting our messages, there is also no need for buffering them
    // process them right away and delete rdkafkamsg and cmsg afterwards
    if (!isCounting()) {
        bool suc = handler->handle(cmsg);
        delete cmsg;
        return suc;
    }

    bool suc = handler->bufferConsumedMessage(cmsg);

    std::string topic = cmsg->topic;
    // std::string topic = getExpectation(cmsg.topic);
    if (Consumer::countingReceiveCounter.count(topic) == 0)
        Consumer::countingReceiveCounter[topic] = 0;
    Consumer::countingReceiveCounter[topic]++;

    if (timeSync) {
        timeSync->tick();
        KDBGCB("ticking");
    }

    return suc;
}

std::string Consumer::getExpectation(std::string topic) {
    std::string t = "";
    if (std::find(Consumer::countingExpectedTopics.begin(), Consumer::countingExpectedTopics.end(), topic) != Consumer::countingExpectedTopics.end()) {
        t = topic;
    } else {
        for (std::string pattern : Consumer::countingExpectedPatterns) {
            if (std::regex_match(topic, std::regex(pattern))) {
                KDBGCB("yes! (pattern " << pattern << " matches)");
                t = pattern;
                break;
            }
        }
    }
    KDBGCB("Expectation for " << topic << " is " << t);
    return t;
}

bool Consumer::isExpectedTopic(std::string topic) {
    KDBGCB("is " << topic << " expected?");
    if (std::find(Consumer::countingExpectedTopics.begin(), Consumer::countingExpectedTopics.end(), topic) != Consumer::countingExpectedTopics.end()) {
        KDBGCB("yes!");
        return true;
    }
    for (std::string pattern : Consumer::countingExpectedPatterns) {
        if (std::regex_match(topic, std::regex(pattern))) {
            KDBGCB("yes! (pattern " << pattern << " matches)");
            return true;
        }
    }
    KDBGCB("no!");
    return false;
}
